dali/operators/audio/resample.h

```diff
@@ -0,0 +1,128 @@
+    if (!in_rate_.HasValue() && !scale_.HasValue() && !out_length_.HasValue())
+      DALI_FAIL("No resampling factor specified! Please supply either the scale, "
+        "the output length or the input and output sampling rates.");
+    quality_ = spec_.template GetArgument<float>("quality");
+    DALI_ENFORCE(quality_ >= 0 && quality_ <= 100, make_string("``quality`` out of range: ",
+      quality_, "\nValid range is [0..100]."));
+    if (spec_.TryGetArgument(dtype_, "dtype")) {
+      // silence useless warning -----------------------------vvvvvvvvvvvvvvvv
+      TYPE_SWITCH(dtype_, type2id, T, (AUDIO_RESAMPLE_TYPES), (T x; (void)x;),
+      (DALI_FAIL(make_string("Unsupported output type: ", dtype_,
+                           "\nSupported types are : ", ListTypeNames<AUDIO_RESAMPLE_TYPES>()))));
+    }
+  }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &outputs, const workspace_t<Backend> &ws) override {
+    outputs.resize(1);
+    if (dtype_ == DALI_NO_TYPE)
+      dtype_ = ws.template Input<Backend>(0).type();
+
+    outputs[0].type = dtype_;
+    CalculateScaleAndShape(outputs[0].shape, ws);
+
+    return true;
+  }
+
+  void CalculateScaleAndShape(TensorListShape<> &out_shape, const workspace_t<Backend> &ws) {
+    const auto &input = ws.template Input<Backend>(0);
+    const TensorListShape<> &shape = input.shape();
+    DALI_ENFORCE(shape.sample_dim() == 1 || shape.sample_dim() == 2,
+      "Audio resampling supports only time series data, with an optional innermost "
+      "channel dimension.");
+    out_shape = shape;
+    int N = shape.num_samples();
+    scales_.resize(N);
+    if (in_rate_.HasValue()) {
+      assert(out_rate_.HasValue());
+      in_rate_.Acquire(spec_, ws, N);
+      out_rate_.Acquire(spec_, ws, N);
+      for (int s = 0; s < N; s++) {
+        double in_rate = in_rate_[s].data[0];
+        double out_rate = out_rate_[s].data[0];
+        scales_[s] = out_rate / in_rate;
+        int64_t in_length = shape.tensor_shape_span(s)[0];
+        int64_t out_length =
+          kernels::signal::resampling::resampled_length(in_length, in_rate, out_rate);
+        out_shape.tensor_shape_span(s)[0] = out_length;
+      }
+    } else if (scale_.HasValue()) {
+      scale_.Acquire(spec_, ws, N);
+      for (int s = 0; s < N; s++) {
+        double scale = scale_[s].data[0];
+        scales_[s] = scale;
+        int64_t in_length = shape.tensor_shape_span(s)[0];
+        int64_t out_length =
+          kernels::signal::resampling::resampled_length(in_length, 1, scale);
+        out_shape.tensor_shape_span(s)[0] = out_length;
+      }
+    } else if (out_length_.HasValue()) {
+      out_length_.Acquire(spec_, ws, N);
+      for (int s = 0; s < N; s++) {
+        int64_t in_length = shape.tensor_shape_span(s)[0];
+        int64_t out_length = out_length_[s].data[0];
+        scales_[s] = 1.0 * out_length / in_length;
```

Comment:
TODO: Verify that the input is not empty if output isn't. Add special case for 0/0.